### ISSUE TITLE
Big slow down when upgrading from 4.10 to 4.11

### ISSUE THREAD

**Issue by giltho** [ocaml/ocaml#10092]
Hi !

## Description

My project is twice as slow using ocaml 4.11 compared to ocaml 4.10, while changing 0 lines in the code.

It is a bit performance sensitive (low times are important when we try to defend our work..).
In particular, one run that is important for us now takes around 62s instead of 30s.

The issue is, the project is quite big, some of it is legacy code...  I have no idea what feature is causing this.
If you have any pointers as to what could cause this, I would very gladly help in trying to track it down.

I don't mind staying on 4.10 (I'm a bit sad that I can't use memtrace but appart from that it's not an issue).
But I'm guessing this is something that is interesting to investigate.

If needed, here's the link to the project (and the specific commit where I revert the upgrade of ocaml).
https://github.com/GillianPlatform/Gillian/tree/gillian-c-2-with-amazon-and-invariants

*EDIT: The following was added on request. Really sorry I didn't give it earlier, I should have.*

## Configuration

I just realised I forgot to add my configuration so here it is:
- macOS Big Sur 11.0.1
- MacBook Pro (16-inch 2016)
- CPU Intel Core i9, 8 cores
- 16Gb 2667MHz DDR4 ram

### Installing esy, dependencies and building the project
 Steps required to build the project. npm is required, since the project's dependencies are at managed by esy.
```sh
$ npm install -g esy@latest #first step is installing esy
$ esy # should install the dependencies and build the project
```
This will take quite a while as it's building ocaml and Z3...

### Running the encryption header verification

Then, the execution I use to mesure the time, because it's quite long it's easier to see the difference :
```sh
# from the root folder of the project
$ esy x gillian-js verify Gillian-JS/Examples/Amazon/deserialize_factory.js -l disabled --no-lemma-proof --amazon
```
For the story, this is verifying some bit of code from Amazon's AWS Encryption SDK.

### Changing the ocaml version

You can change the ocaml version by going into the `esy.json` file at the root and replacing both occurences of
`4.10.x` with `4.11.x` (the x shouldn't be substituted but actually written like that in the file so that `esy` can simply find the latest patch.

Then you can rebuild using
```sh
$ esy
```
This will unfortunately rebuild all dependencies (including ocaml).

### Uninstalling esy
If you're not an esy user, you might want to clean up after you're done, because esy keeps every version of the ocaml compiler it ever needed and that takes some space.
```
$ rm -rf ~/.esy
$ npm uninstall -g esy
```

## Additional information - Exploring the project

The project is composed of several components:

The GillianCore folder contains the `gillian` library. It mostly exposes some big functors that helps building symbolic static analysis tools.  In particular, the CLI folder builds a uniform cli to analyse programs.

The parameters to that functor are modules that represent an "instantiation" of Gillian. One instantiates Gillian to analyse programs written in a specific language. One has to plug a memory model into Gillian to be able to perform analysis. An instantiation is composed of 4 module : a concrete memory model, a symbolic memory model (both are modules following the right interface), a "parser and compiler" module that describes how to compile a program in your language into our intermediate representation (`GIL`, described in `Gil_syntax`) and a module called `External` that is barely used.

In the repo, you'll find 3 instantiations of Gillian:
- `wisl`, a small toy language we use to teach / understand things
- `Gillian-C`, an instantiation of Gillian for the C language, using bits of `CompCert` as compiler.
- `Gillian-JS`, an instantiation of Gillian for JavaScript, inherited from previous work.

For the execution I described above, one only needs to understand Gillian-JS.
The root of the program is the file `Gillian-JS/bin/gillian_js.ml`, which uses the CLI functor and calls its `main` function.
From that you should be able to follow execution if needed..

About the dependencies, the field `"dependencies"` of `esy.json` describes them, but note that Batteries and Compcert are only used in Gillian-C, and therefore can't be the linked to the issue.

### Some more commands you can try

```sh
$ esy x gillian-js verify Gillian-JS/Examples/JaVerT/BST.js -l disabled --no-lemma-proof  
$ esy x gillian-c verify Gillian-C/examples/verification/sll.c -l disabled
```

### Further information

That is really all I know. I can help when it comes to exploring the codebase more if needed, but I really have no clue of where to start looking for what causes the slow down.

**Comment by dra27** [comment]
2 first steps to help - the project is large, so can the issue at least have precise instructions to build and show what you know now? Additionally, are you able to try it with the OCaml 4.12 alpha3 release?

**Comment by giltho** [comment]
I added some details, sorry I should have added more beforehand

I've also tried building with 4.12 alpha3 but there are a lot of dependencies that are conflicting. I'm gonna try to see if I can find a commit for each of the dependencies that are conflicting but I'm not sure it's possible

EDIT: trying to install ppxlib, I'm running into some issue related to this https://github.com/ocaml/dune/issues/3766, and I don't know how to work around it, so I don't think I'll be able to try with 4.12...
I get some `"stdlib_shims.a" not available` when esy is running `dune build -p ppxlib`

**Comment by gasche** [comment]
I spent some time yesterday trying to understand the issue. I don't have a good answer, but here are some things I noticed (not very coherently, sorry about that):
- it looks like your tests are coherently twice slower, it is not just one of the lemmas or specifications that is taking much longer
- I had a hard time getting a good performance profile; in particular `perf record --call-graph=dwarf ... perf report` gives fairly bad traces where essentially only OCaml runtime functions show up, and almost none of the user-code functions, and I don't know why that is. (At first I suspected that the code fails to be compiled with `-g`, possibly (but I'm not sure) because of the fact that the `(flags ...)` directives used to pass `-open` flags do not include `:standard`. I tried various things to ensure that `-g` is used, and I think it is, but the traces are still pretty bad.)
- In the performance profiles I got, one striking thing is that a non-neglectible amount of time is spent in `camlinternalOO` in the 4.11 profile, while they do not show up in the 4.10 profile. If the regression comes from OCaml, a possible candidate would be a regression in method caching or some other sort of object-oriented-related compilation change. (None of the 4.11 changes jumps to mind as a probable culprit.)
- There are some places that look like they could be easily improved to make the code faster. In particular, profiles show a fair amount of time spent in `Reduction.reduce_assertion_loop` (iirc.) that is a map function that often behaves as the identity, with a rather naive fixpoint-computation check (it should be `if not (a == result) && a <> result`, but is incorrectly implemented as `if a <> result && not (a == result)`) that could be made sensibly faster by having a mutable boolean flag to indicate when a change occurred deep in the function.
- The use of `esy` seemed convenient at first (it's really easy to install all dependencies and build in just one command), but in fact it makes debugging for compiler-related changes very difficult because I have no idea how to test the code on another OCaml version that the one that is specified in `esy.json`. I guess this is the usual downside of "locked" dependencies, you can't easily change things. I would naturally try to `bisect` this issue over intermediate compiler versions between 4.10 and 4.11, but I really don't know how to do this with an `esy`-based build system, while standard dune+opam would be relatively easy to manage. I tried to do a standard dune+opam build of the system, but it uses features of esy (specifying things that are not package in opam by their github-project + commit-hash, specifically for JS_parser) that I wasn't sure how to avoid. If you want us to look further into the issue, I think it would be very helpful if you could do the work of simplifying the build system yourself, by vendoring the dependencies that are not package on opam in such a way that "dune" can find them without using esy.

**Comment by giltho** [comment]
Thank you so much for taking time to look into this !

First of all, thanks for for the advice regarding reduction ! Will modify this asap.

I'm afraid there's no easy solution to bisect the version of the compiler using esy right now. I will work on a version of Gillian that uses vendored dependencies and get back to you. I would love to have the dune+opam workflow available anyway, but I can't do it without vendoring right now (merging some PRs such as https://github.com/flowtype/ocaml-ppx_gen_rec/pull/5 and https://github.com/ocaml-community/sedlex/pull/94 or a release of visitors would be required...).

Regarding the flags, I will also add the `:standards` flags, where needed, but I get precise backtraces of in case of failure, so I think the `-g` flag is already enabled.

What would be your process to bisect the issue over intermediate versions of the compiler ? Does opam have a plugin for that ? I could do that myself so you don't spend too much time on it...

**Comment by avsm** [comment]
You can pin to a particular hash in opam by using the `pin-depends` field in your package; see https://opam.ocaml.org/doc/Manual.html#opamfield-pin-depends (or https://github.com/ocamllabs/opam-monorepo/blob/master/opam-monorepo.opam.locked for an example of a larger usage of this)

**Comment by giltho** [comment]
@avsm sorry to ping you, I don't really know how to pin a specific commit of the ocaml compiler with opam
I've tried 
```
[ "ocaml.4.11.0+beta1" "git+https://github.com/ocaml/ocaml.git#<commit_hash>"]
```
but I get
```
[WARNING] Couldn't retrieve opam file from versioned source, using the one found locally.
```
What should be my pin-depends here ?

Thanks !

**Comment by dbuenzli** [comment]
@giltho maybe you will have better luck with [these instructions](https://github.com/ocaml/ocaml/blob/trunk/HACKING.adoc#testing-with-opam).

**Comment by giltho** [comment]
aaaah instructions explain that it's ocaml-variants that needs to be pinned.
Thank you !

**Comment by giltho** [comment]
I tried bisecting automatically and let the script run all night since every test takes ~15mn (compiling ocaml + z3...) but most commits fail since batteries can't compile and I don't know how to fix it.

Maybe you'll have better luck.
FIrst of all in the Gillian repo there's a branch that can be used with opam fully :https://github.com/GillianPlatform/Gillian/tree/bisect-ocaml-compiler

There's a trick for z3 because of this issue https://github.com/Z3Prover/z3/issues/2305, the opam version isn't working I had to export my own and opam won't let me use `setenv` to export that variable automatically.

## Instructions
```
git clone https://github.com/GillianPlatform/Gillian.git
git checkout bisect-ocaml-compiler
git submodule update --init
```

then from the `ocaml` folder, I used the following `zsh` script to bisect : 

```zsh
#!/bin/zsh

COMMIT_HASH=$(git rev-parse HEAD)
GILLIAN_PATH=/YOUR/PATH/TO/Gillian

function switch_back () {
  opam switch 4.11.1
  opam switch remove $COMMIT_HASH -y
}

opam switch create $COMMIT_HASH --empty
eval $(opam env)
opam install . -y --working-dir

cd $GILLIAN_PATH

echo "INSTALLING OPAM DEPS"
opam install . -y --deps-only --unlock-base


RUNTIME_PATH=$GILLIAN_PATH/_build/install/default/share/gillian-js
Z3LIB=~/.opam/$COMMIT_HASH/lib/z3


echo "COMPUTING AMAZON TIME"
TIME=$(DYLD_LIBRARY_PATH=$Z3LIB LD_LIBRARY_PATH=$Z3LIB GILLIAN_JS_RUNTIME_PATH=$RUNTIME_PATH opam exec -- dune exec Gillian-JS/bin/gillian_js.exe -- verify Gillian-JS/Examples/Amazon/deserialize_factory.js -l disabled --no-lemma-proof --amazon | tail -1 | awk -F ": " '{print $2}') || exit 125

echo "TIME IS: $TIME"

re='^[0-9]+[.][0-9]+'
if ! [[ $TIME =~ $re ]]; then
  switch_back
  exit 125
fi

if (( $TIME > 45 )); then
  switch_back
  exit 1
fi

switch_back
```

**Comment by gasche** [comment]
Thanks! I have not tried this yet, but I tried the following things (from the idea that the regression may be object-method-dispatch related):
- I tried a microbenchmark of very simple method dispatch (with constant classes and parametrized classes, with and without inheritance)
- I tried a simple microbenchmark of `fold` instances generated by ppx.visitors

Neither appear to differ significantly between 4.10 and 4.11 in my testing.

For reference, my "visitors" microbenchmark was as follows:
```
type expr =
| EConst of int
| EAdd of expr * expr
[@@deriving visitors { variety = "fold" }]

let sum expr = 
  let visitor = object
    inherit [_] fold as super
    method visit_int _env n = n
    method build_EConst _env n = n
    method build_EAdd _env a b = a + b
  end
  in visitor#visit_expr () expr

let rec large n =
  if n = 0 then EConst 1
  else
    let expr = large (n - 1) in
    EAdd (expr, expr)

let example = large (read_int ())

let () = Printf.printf "%d\n%!" (sum example)
```

**Comment by gasche** [comment]
I managed to bisect the code thanks to your simpler build system (I used essentially the same approach as you did, manually fixing Batteries (and cherry-picking a topkg fix) on revisions that required it). I have identified the culprit for the regression, which is #9096 (cc @stedolan), and this is indeed related to object-oriented compilation. I propose a fix in #10096 (against `trunk`, but applies on 4.11 just fine).

**Comment by gasche** [comment]
(Note: To me this suggests that Gillian may improve its performance significantly by implementing substitutions "manually" instead of relying on method dispatch -- assuming substitutions (SSubst and SESubst) are indeed the only uses of object-oriented code in your critical sections.)

**Comment by bluddy** [comment]
Sorry to butt in @gasche. Is there an explanation of method caching in OCaml? I haven't found one anywhere, including in the compiler docs.

My guess is that since method dispatch in OCaml requires a dynamic hashtable lookup, it's expensive, and we cache the most recent methods called at runtime. Is this correct or is it something else?

**Comment by gasche** [comment]
@bluddy I don't know much about this myself, my impression is that there are different sorts of caching involved, but I don't know where it would be documented. (I would try asking Jacques Garrigue directly for a description of these mechanisms.)

There seems to be some caching of method dispatch (note: the bytecode interpreter source suggests that method resolution is performed by dichotomic search rather than hashtable lookup in many case), but the one that is involved in this particular regression is different, it is a caching of the construction of the class tables at class-declaration time. If you declare a new class (or object) from within a function, the compiler is careful to initialize the class/object tables lazily, using (hidden) global mutable state, instead of doing it on every call to the function.

**Comment by garrigue** [comment]
Indeed, method call site caching is implemented by keeping the index in the vtable of the last method call for each call site. It is then cheap to check whether this index is valid on the next call. Otherwise, one recomputes the index through dichotomy search. This is not documented, as this is just a compiler optimization. Note that the dichotomy search itself is not that expensive either. Of course, I don't see how a change in internal location representation would break this cacheing.

@gasche, you suggest that the problem is rather in the cacheing mechanism use for vtable construction. Can you be more precise. Howe would a change in location representation break it?

**Comment by gasche** [comment]
@garrigue I think that this part is more easily discussed in #10096, my PR which has the fix. To answer your question, the change in location representation came with some "smart constructors" on `Tfunction` nodes to deal with currification (I guess), which were buggy in the sense that they removed a call to `Translobj.oo_wrap` in some cases.

**Comment by gasche** [comment]
The fix is now upstream in the compiler codebase. It will be part of the upcoming 4.12 release (a release is planned for mid-January). I also asked for an extra 4.11 bugfix release as a delayed Christmas present, but I don't know how excited our release manager @Octachron is about the idea. Closing this; thanks a lot @giltho for the report, and the help working on the issue.

**Comment by dra27** [comment]
On 26 Dec 2020, at 15:52, Gabriel Scherer <notifications@github.com> wrote:
The fix is now upstream in the compiler codebase. It will be part of the upcoming 4.12 release (a release is planned for mid-January). I also asked for an extra 4.11 bugfix release as a delayed Christmas present, but I don't know how excited our release manager @Octachron<https://github.com/Octachron> is about the idea.

Happy Christmas to you both!

Just a small addition to this that when we included the mingw64 fixes in 4.10.2, I was assuming we’d do a 4.11.2 release as it would be odd to have fixed something in the 4.10 series and then not the 4.11.

Best,


David

**Comment by giltho** [comment]
I was on vacation so I wasn't very active in the end, but I just wanted to add a small comment to say thank you for being so efficient on fixing this ! 
(and Happy New year 🙂)
